src/wallet/rpcdump.cpp

```diff
@@ -840,195 +920,144 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Pubkey is not a valid public key");
             }
-
-            // Now set script to the redeemScript so we parse the inner script as P2WSH or P2WPKH below
-            script = redeemScript;
-            ExtractDestination(script, dest);
+            import_data.pubkeys.emplace(pubkey.GetID(), pubkey);
         }
-
-        // (P2SH-)P2WSH
-        if (!witness_script_hex.empty() && script.IsPayToWitnessScriptHash()) {
-            if (!IsHex(witness_script_hex)) {
-                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid witness script: must be hex string");
+        for (size_t i = 0; i < keys.size(); ++i) {
+            const auto& str = keys[i].get_str();
+            CKey key = DecodeSecret(str);
+            if (!key.IsValid()) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid private key encoding");
             }
-
-            // Generate the scripts
-            std::vector<unsigned char> witness_script_parsed(ParseHex(witness_script_hex));
-            CScript witness_script = CScript(witness_script_parsed.begin(), witness_script_parsed.end());
-            CScriptID witness_id(witness_script);
-
-            // Check that the witnessScript and scriptPubKey match
-            if (GetScriptForDestination(WitnessV0ScriptHash(witness_script)) != script) {
-                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "The witnessScript does not match the scriptPubKey or redeemScript");
+            CPubKey pubkey = key.GetPubKey();
+            CKeyID id = pubkey.GetID();
+            if (import_data.pubkeys.count(id)) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Both public and private key specified for " + HexStr(pubkey.begin(), pubkey.end()));
             }
+            import_data.privkeys.emplace(id, key);
+        }
 
-            // Add the witness script as watch only only if it is not for P2SH-P2WSH
-            if (!scriptpubkey_script.IsPayToScriptHash() && !pwallet->AddWatchOnly(witness_script, timestamp)) {
-                throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
-            }
+        // Watchonly and private keys
+        if (watchOnly && import_data.privkeys.size()) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Watch-only addresses should not include private keys");
+        }
+        if (watchOnly && (import_data.pubkeys.size() || import_data.redeemscript || import_data.witnessscript)) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Watch-only addresses should not include any keys or scripts");
+        }
 
-            if (!pwallet->HaveCScript(witness_id) && !pwallet->AddCScript(witness_script)) {
-                throw JSONRPCError(RPC_WALLET_ERROR, "Error adding p2wsh witnessScript to wallet");
-            }
+        // TODO: it seems we're not actually requiring watchonly to be set for non-solvable inputs, fix this
```

Comment:
That's a great idea. I've implemented that and taken it a bit further. I think all new error conditions (that don't signify seriously broken things like P2SH-inside-P2SH) are now reported as warnings instead of causing the RPC to fail.

Perhaps we want to introduce a new mode where all these warnings become errors, though `-deprecatedrpc`, but let's discuss that in a separate RPC.